File: sort_images.py
import os
import re
from PIL import Image

def get_colorfulness(image_path):
    try:
        with Image.open(image_path) as img:
            img = img.convert('RGB')
            img.thumbnail((100, 100)) # Resize for speed
            
            pixels = list(img.getdata())
            total_diff = 0
            for r, g, b in pixels:
                max_val = max(r, g, b)
                min_val = min(r, g, b)
                total_diff += max_val - min_val
                
            avg_diff = total_diff / len(pixels)
            return avg_diff
    except Exception as e:
        print(f"Error processing {image_path}: {e}")
        return 9999 # Put errors at the end

# Extract current list from script.js
with open('script.js', 'r', encoding='utf-8') as f:
    content = f.read()

# Find the rawImages array
match = re.search(r'const rawImages = \[\s*([\s\S]*?)\s*\];', content)
if not match:
    print("Could not find rawImages array in script.js")
    exit(1)

array_content = match.group(1)
# Parse individual paths (ignoring quotes and commas)
paths = re.findall(r"'([^']+)'", array_content)

# Calculate colorfulness for each
print(f"Calculating colorfulness for {len(paths)} images...")
scored_paths = []
for p in paths:
    full_path = os.path.join('.', p) # Assuming paths are relative to current dir
    score = get_colorfulness(full_path)
    scored_paths.append((p, score))

# Sort by colorfulness (low to high)
scored_paths.sort(key=lambda x: x[1])

print("Top 5 least colorful:")
for p, s in scored_paths[:5]:
    print(f"{s:.2f}: {p}")

print("\nTop 5 most colorful:")
for p, s in scored_paths[-5:]:
    print(f"{s:.2f}: {p}")

# Generate new array JS
new_array_content = ',\n'.join([f"        '{p}'" for p, s in scored_paths])
new_raw_images = f"const rawImages = [\n{new_array_content}\n    ];"

# Replace in script.js
new_content = content[:match.start()] + new_raw_images + content[match.end():]
with open('script.js', 'w', encoding='utf-8') as f:
    f.write(new_content)

print("\nscript.js updated successfully!")

File: sort_images.swift
import Foundation
import AppKit

let fileManager = FileManager.default
let currentPath = fileManager.currentDirectoryPath

func getColorfulness(imagePath: String) -> Double {
    guard let image = NSImage(contentsOfFile: imagePath),
          let tiffData = image.tiffRepresentation,
          let bitmap = NSBitmapImageRep(data: tiffData) else {
        return 9999.0
    }
    
    // Calculate colorfulness (average max-min diff of RGB)
    let width = bitmap.pixelsWide
    let height = bitmap.pixelsHigh
    let bytesPerRow = bitmap.bytesPerRow
    let samplesPerPixel = bitmap.samplesPerPixel
    guard let bitmapData = bitmap.bitmapData else { return 9999.0 }
    
    var totalDiff: Double = 0
    var count: Double = 0
    
    // Sample every 4th pixel to speed up
    for y in stride(from: 0, to: height, by: 4) {
        for x in stride(from: 0, to: width, by: 4) {
            let offset = y * bytesPerRow + x * samplesPerPixel
            // Ensure we don't go out of bounds
            if offset + 2 < bytesPerRow * height {
                let r = Double(bitmapData[offset])
                let g = Double(bitmapData[offset + 1])
                let b = Double(bitmapData[offset + 2])
                
                let maxVal = max(r, max(g, b))
                let minVal = min(r, min(g, b))
                totalDiff += (maxVal - minVal)
                count += 1
            }
        }
    }
    
    if count == 0 { return 9999.0 }
    return totalDiff / count
}

do {
    let scriptPath = currentPath + "/script.js"
    let content = try String(contentsOfFile: scriptPath, encoding: .utf8)
    
    // Find the rawImages array using regex
    let regex = try NSRegularExpression(pattern: "const rawImages = \\[(.*?)\\];", options: [.dotMatchesLineSeparators])
    let nsRange = NSRange(content.startIndex..<content.endIndex, in: content)
    
    if let match = regex.firstMatch(in: content, options: [], range: nsRange) {
        let arrayRange = match.range(at: 1)
        if let swiftRange = Range(arrayRange, in: content) {
            let arrayContent = String(content[swiftRange])
            
            // Extract paths
            let pathRegex = try NSRegularExpression(pattern: "'([^']+)'")
            let pathMatches = pathRegex.matches(in: arrayContent, options: [], range: NSRange(arrayContent.startIndex..<arrayContent.endIndex, in: arrayContent))
            
            var paths: [String] = []
            for m in pathMatches {
                if let r = Range(m.range(at: 1), in: arrayContent) {
                    paths.append(String(arrayContent[r]))
                }
            }
            
            print("Calculating colorfulness for \(paths.count) images...")
            var scoredPaths: [(String, Double)] = []
            for p in paths {
                let fullPath = currentPath + "/" + p
                let score = getColorfulness(imagePath: fullPath)
                scoredPaths.append((p, score))
            }
            
            // Sort by colorfulness
            scoredPaths.sort { $0.1 < $1.1 }
            
            print("Top 5 least colorful:")
            for i in 0..<min(5, scoredPaths.count) {
                print(String(format: "%.2f: %@", scoredPaths[i].1, scoredPaths[i].0))
            }
            
            print("Top 5 most colorful:")
            for i in max(0, scoredPaths.count - 5)..<scoredPaths.count {
                print(String(format: "%.2f: %@", scoredPaths[i].1, scoredPaths[i].0))
            }
            
            // Generate new array content
            let newArrayContent = scoredPaths.map { "        '\($0.0)'" }.joined(separator: ",\n")
            let newRawImages = "const rawImages = [\n\(newArrayContent)\n    ];"
            
            let fullMatchRange = match.range
            if let fullSwiftRange = Range(fullMatchRange, in: content) {
                var newContent = content
                newContent.replaceSubrange(fullSwiftRange, with: newRawImages)
                
                try newContent.write(toFile: scriptPath, atomically: true, encoding: .utf8)
                print("script.js updated successfully!")
            }
        }
    }
} catch {
    print("Error: \(error)")
}
